sdk/go/pulumi/errors.go

```diff
@@ -0,0 +1,40 @@
+// Copyright 2016-2024, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pulumi
+
+import (
+	pulumirpc "github.com/pulumi/pulumi/sdk/v3/proto/go"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type InputPropertyErrorDetails struct {
+	PropertyPath string
+	Reason       string
+}
+
+func NewInputPropertiesError(message string, errors ...InputPropertyErrorDetails) error {
+	// Convert the errors to a slice of proto messages.
+	errorDetails := pulumirpc.InputPropertiesError{}
+	for _, e := range errors {
+		errorDetails.Errors = append(errorDetails.Errors, &pulumirpc.InputPropertiesError_PropertyError{
+			PropertyPath: e.PropertyPath,
+			Reason:       e.Reason,
+		})
+	}
+
+	s, _ := status.Newf(codes.InvalidArgument, "%s", message).WithDetails(&errorDetails)
+	return s.Err()
+}
```

Comment:
I don't mind how we land on this, but I feel like we should not abuse the fact that we have the native gRPC types available to us here. I'd support us defining our own error types e.g. 

```go
type InputPropertiesError struct {
  Err error
  Details []InputPropertyErrorDetails
}
```

and looking for these in the server and translating appropriately. I don't think it'll cost us hugely not to separate concerns like this, but e.g. if we do I think it makes some things neater -- e.g. customers who want to test component providers or the like can assert that `errors.As(..., pulumi.InputPropertiesError, ...)` or the like without having to get down and dirty with gRPC.

Reply:
Yeah that makes sense.  I'm happy to have a separate error type here.  I would make it 

```
type InputPropertiesError struct {
  Message string
  Errors []InputPropertyErrorDetails
}
```

though for consistency with other languages :thinking: I think that also matches the interface you're proposing below.